crates/ide_assists/src/handlers/generate_function.rs

```diff
@@ -132,7 +135,43 @@ impl FunctionBuilder {
         let fn_name = fn_name(&path)?;
         let (type_params, params) = fn_args(ctx, target_module, &call)?;
 
-        Some(Self { target, fn_name, type_params, params, file, needs_pub })
+        // should_render_snippet intends to express a rough level of confidence about
+        // the correctness of the return type.
+        //
+        // If we are able to infer some return type, and that return type is not unit, we
+        // don't want to render the snippet. The assumption here is in this situation the
+        // return type is just as likely to be correct as any other part of the generated
+        // function.
+        //
+        // In the case where the return type is inferred as unit it is likely that the
+        // user does in fact intend for this generated function to return some non unit
+        // type, but that the current state of their code doesn't allow that return type
+        // to be accurately inferred.
+        let (ret_type, should_render_snippet) = {
+            if let Some(ty) = ctx.sema.type_of_expr(&ast::Expr::CallExpr(call.clone())) {
+                if ty.is_unknown() || ty.is_unit() {
+                    (make::ret_type(make::ty_unit()), true)
+                } else if let Ok(rendered) = ty.display_source_code(ctx.db(), target_module.into())
+                {
+                    (make::ret_type(make::ty(&rendered)), false)
+                } else {
+                    (make::ret_type(make::ty_unit()), true)
+                }
+            } else {
+                (make::ret_type(make::ty_unit()), true)
+            }
```

Comment:
I'm not too happy with all the branching here. Definitely open to suggestions for cleaning this up!

Reply:
I'd write it like this:

```rust
        let (ret_ty, should_render_snippet) = {
            match ctx.sema.type_of_expr(&ast::Expr::CallExpr(call.clone())) {
                Some(ty) if ty.is_unknown() || ty.is_unit() => (make::ty_unit(), true),
                Some(ty) => {
                    let rendered = ty.display_source_code(ctx.db(), target_module.into());
                    match rendered {
                        Ok(rendered) => (make::ty(&rendered), false),
                        Err(_) => (make::ty_unit(), true),
                    }
                }
                None => (make::ty_unit(), true),
            }
        };
        let ret_type = make::ret_type(ret_ty);
```

Reply:
Thanks for the suggestion - this looks a lot better!